### PR TITLE
refund after sign

### PR DESCRIPTION
**Pull request by ailisp**
Refund all deposit if sign fails.
- If sign call itself fails (does not include the cross contract call invoked by sign, for example incorrect version), near runtime will not take required deposit, does not need to refund
- sign will refund more than required deposit if sign it self fails, then if sign invoked cross contract call fails, it further refund required deposit. As a result all deposit is refunded

Refund more than required deposit if sign success.

Address #708

### CODE REVIEW

File: chain-signatures/contract/src/lib.rs [hunk 1]
```diff
@@ -718,15 +731,28 @@ impl VersionedMpcContract {
     pub fn clear_state_on_finish(
         &mut self,
         request: SignatureRequest,
+        refund_to_on_fail: AccountId,
```
[the remark this answers is not in the document]
Comment (ailisp) [reply chain-signatures/contract/src/lib.rs]: Yes, make it only refund once now

[the remark this answers is not in the document]
Comment (ailisp) [reply chain-signatures/contract/src/lib.rs]: SignatureRequest also used in protocol, in protocol SignatureManager.poke, the refund and requester info isn't relevant in that context, therefore I'll make another ContractSignatureRequest struct enclosing the original SignatureRequest, use specifically for contract call

File: chain-signatures/contract/src/lib.rs [hunk 2]
```diff
@@ -696,7 +692,7 @@ impl VersionedMpcContract {
                 // by it won't execute.
                 let yield_promise = env::promise_yield_create(
                     "clear_state_on_finish",
-                    &serde_json::to_vec(&(&contract_signature_request)).unwrap(),
+                    &serde_json::to_vec(&(&contract_signature_request,)).unwrap(),
```
Comment (ailisp) [inline chain-signatures/contract/src/lib.rs:695]: Can't believe missing this `,` took me an afternoon to fix 😿

File: chain-signatures/contract/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ailisp) [reply chain-signatures/contract/src/lib.rs]: Done

File: chain-signatures/contract/src/lib.rs [hunk 3]
```diff
@@ -677,12 +683,16 @@ impl VersionedMpcContract {
     }
 
     #[private]
-    pub fn sign_helper(&mut self, request: SignatureRequest) {
+    pub fn sign_helper(&mut self, contract_signature_request: ContractSignatureRequest) {
         match self {
             Self::V0(mpc_contract) => {
+                // refund must happen in clear_state_on_finish, because regardless of this success or fail
```
Comment (volovyks) [inline chain-signatures/contract/src/lib.rs:689]: Should we move this comment to where the refund is actually happening?

**Review by volovyks — DISMISSED** [review]

File: chain-signatures/contract/src/lib.rs (hunk 3 above)
Comment (ailisp) [reply chain-signatures/contract/src/lib.rs]: OK

File: chain-signatures/contract/tests/sign.rs [hunk 4]
```diff
@@ -57,6 +59,124 @@ async fn test_contract_sign_request() -> anyhow::Result<()> {
...
+
+    let new_balance = alice.view_account().await?.balance;
+    println!(
+        "{} {} {}",
+        balance.as_millinear(),
+        new_balance.as_millinear(),
+        contract.view_account().await?.balance
+    );
+    assert!(
+        balance.as_millinear() - new_balance.as_millinear() < 10,
```
Comment (volovyks) [inline chain-signatures/contract/tests/sign.rs:173]: Sorry for the late comment. Can we also ensure that our balance is not lower than expected?

Comment (volovyks) [reply chain-signatures/contract/tests/sign.rs:173]: I'm talking about mpc-contract balance

Comment (ailisp) [reply chain-signatures/contract/tests/sign.rs:173]: This is implicitly ensured, i tested manually an intentional over refund, which cause this line underflow and fail. I can make the check explicit

**Comment by ailisp** [comment]
@volovyks ptal

**Review by ppca — APPROVED** [review]
LGTM

**Pull request merged**